torch/_functorch/_activation_checkpointing/ac_reorder_inference_pass.py

```diff
@@ -0,0 +1,158 @@
+        impure_random = torch._inductor.config.fallback_random
+    return node.is_impure(impure_random)
+
+
+def _is_backward_node(node: fx.Node) -> bool:
+    """Check if node is in backward region via annotation"""
+    return node.meta.get("custom", {}).get("backward") is not None
+
+
+def rematerialize_nodes_with_ac_annotations(gm: fx.GraphModule) -> fx.GraphModule:
+    """
+    Duplicate checkpointed nodes for backward use. DCE removes unused forward versions.
+    """
+    # Early exit if no checkpointed ops
+    if not has_recomputable_ops(gm):
+        return gm
+
+    # Ban RNG ops in checkpointed regions (like partitioner does for inference mode)
+    # Note: RNG functionalization is not currently supported for inference mode graphs
+    if has_recomputable_rng_ops(gm):
+        raise RuntimeError(
+            "Activation checkpoint reordering in fullgraph does not support RNG ops in checkpointed regions. "
+            "RNG state cannot be properly synchronized between "
+            "forward and recomputed backward operations. Please move RNG operations outside "
+            "of checkpoint regions, or use joint graph mode (where partitioner handles RNG)."
+        )
+
+    # Apply partitioner's edge case handling FIRST
+    # This marks nodes that must be saved (MUST_SAVE, collectives, mutations, etc.)
+    gm = cleanup_recompute_tags(gm, is_default_partition=False)
+
+    if not config.unsafe_allow_optimization_of_collectives:
+        force_save_collectives(gm)
+
+    force_save_bw_mutation_src(gm)
+
+    # Find backward boundary and build ordering
+    first_node_in_bwd = None
+    bwd_start = None
+    order = {}
+    for idx, node in enumerate(gm.graph.nodes):
+        order[node] = idx
+        if _is_backward_node(node) and first_node_in_bwd is None:
+            first_node_in_bwd = node
+            bwd_start = idx
+
+    if first_node_in_bwd is None:
+        return gm
+
+    # Build reordered graph
+    new_graph = fx.Graph()
+    env: dict[fx.Node, fx.Node] = {}
+    recomputed_nodes: dict[fx.Node, fx.Node] = {}
+
+    # Add placeholders
+    for node in gm.graph.find_nodes(op="placeholder"):
+        env[node] = new_graph.node_copy(node, lambda x: env[x])
+
+    def gather_checkpointed_deps(node: fx.Node, visited: set) -> None:
+        if node in visited:
+            return
+        visited.add(node)
+        for inp in node.all_input_nodes:
+            if inp.op != "placeholder" and must_recompute(inp):
+                gather_checkpointed_deps(inp, visited)
+
+    # Insert forward nodes
+    for node in list(gm.graph.nodes)[:bwd_start]:
+        if node.op == "placeholder":
+            continue
+        env[node] = new_graph.node_copy(node, lambda x: env[x])
+
+    # Insert backward nodes
+    for node in list(gm.graph.nodes)[bwd_start:]:
+        if node.op == "output":
+            continue
+
+        # Gather ALL checkpointed deps needed by this node
+        deps = set()
```

Comment:
I wonder if it is better if we reuse the same set instance for the entire loop over backward nodes

Reply:
Hmm don't we want to only preserve ordering for specific backward node?

Reply:
or do you see benefit for globally sorting everything first?

Reply:
Yeah you're right. I don't think we want to reuse this set instance.
But what I feel might be happening is that we're regathering a lot of the same deps multiple times, it ends up working out because we filter for "dep not in recomputed_nodes" later, but maybe we can do something better?

I wonder if it might be better to to gather_checkpointed_deps have already check "dep not in recomputed_nodes" instead, so that deps should be disjoint.

Reply:
Yeah soiunds good! I will check in the gather step and assert later